language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -129,45 +129,45 @@ typeCheckStep = do
         else modify (subtract 1)
 
 -- | Extract kind information from a type.
-kindOf :: Type TyNameWithKind a -> TypeCheckM a (Kind ())
-kindOf TyInt{} = pure (Size ())
+kindOf :: Type TyNameWithKind a -> TypeCheckM a (Kind a)
+kindOf (TyInt x _) = pure (Size x)
 kindOf (TyFun x ty' ty'') = do
     k <- kindOf ty'
     k' <- kindOf ty''
     if isType k && isType k'
-        then pure (Type ())
+        then pure (Type x)
         else
             if isType k
-                then throwError (KindMismatch x (void ty'') k' (Type ()))
-                else throwError (KindMismatch x (void ty') k (Type ()))
+                then throwError (KindMismatch x ty'' k' (Type x))
+                else throwError (KindMismatch x ty' k (Type x))
 kindOf (TyForall x _ _ ty) = do
     k <- kindOf ty
     if isType k
-        then pure (Type ())
-        else throwError (KindMismatch x (void ty) (Type ()) k)
-kindOf (TyLam _ _ k ty) =
-    [ KindArrow () (void k) k' | k' <- kindOf ty ]
-kindOf (TyVar _ (TyNameWithKind (TyName (Name (_, k) _ _)))) = pure (void k)
-kindOf (TyBuiltin _ b) = do
+        then pure (Type x)
+        else throwError (KindMismatch x ty (Type x) k)
+kindOf (TyLam x _ k ty) =
+    [ KindArrow x k k' | k' <- kindOf ty ]
+kindOf (TyVar _ (TyNameWithKind (TyName (Name (_, k) _ _)))) = pure k
+kindOf (TyBuiltin x b) = do
     (BuiltinTable tyst _) <- ask
     case M.lookup b tyst of
-        Just k -> pure k
+        Just k -> pure $ x <$ k
         _      -> throwError InternalError
 kindOf (TyFix x _ ty) = do
     k <- kindOf ty
     if isType k
-        then pure (Type ())
-        else throwError (KindMismatch x (void ty) (Type ()) k)
+        then pure (Type x)
+        else throwError (KindMismatch x ty (Type x) k)
 kindOf (TyApp x ty ty') = do
     k <- kindOf ty
     case k of
         KindArrow _ k' k'' -> do
             k''' <- kindOf ty'
             typeCheckStep
-            if k' == k'''
+            if void k' == void k'''
```

Comment:
The `void`s are due to the fact that the `Eq` instance of `Kind` checks annotation for equality as well. And this also holds for all the errors from `Language.PlutusCore.Error`. We need to do something about that, but not in this PR.